main/lsp/LSPTypechecker.cc

```diff
@@ -334,19 +335,22 @@ LSPTypechecker::SlowPathResult LSPTypechecker::runSlowPath(LSPFileUpdates &updat
     auto slowPathOp = std::make_optional<ShowOperation>(*config, ShowOperation::Kind::SlowPathBlocking);
     Timer timeit(logger, "slow_path");
     ENFORCE(updates.typecheckingPath != TypecheckingPath::Fast || config->disableFastPath);
-    ENFORCE(updates.updatedGS.has_value());
-    if (!updates.updatedGS.has_value()) {
-        Exception::raise("runSlowPath called with an update that lacks an updated global state.");
-    }
     logger->debug("Taking slow path");
 
     ENFORCE(this->cancellationUndoState == nullptr);
     if (cancelable) {
+        auto trackUntyped = this->gs->trackUntyped;
+        auto savedGS = std::exchange(this->gs, payload::copyForSlowPath(*this->gs, this->config->opts));
+
         this->cancellationUndoState =
-            std::make_unique<UndoState>(std::move(gs), std::move(this->indexedFinalGS), updates.epoch);
-    }
+            std::make_unique<UndoState>(std::move(savedGS), std::move(this->indexedFinalGS), updates.epoch);
 
-    this->gs = std::move(updates.updatedGS.value());
+        pipeline::setGlobalStateOptions(*this->gs, this->config->opts);
```

Comment:
This feels out of place to me: my intention with `setGlobalStateOptions` was that it would be called once, and then GlobalState would be the source of truth for mutable things. e.g. the act of copying GlobalState should copy options.